server/src/controllers/christmas-tree/permits.es6

```diff
@@ -165,26 +142,24 @@ at pay.gov ${updatedPermit.paygovError}`
 const recordPayGovError = (error, result, res, permit, requestType) => {
   logger.error(`ERROR: ServerError: Pay.gov- ${error}. Updating permit with error while ${requestType}`);
   return paygov.getResponseError(requestType, result)
-    .then(paygovError => updatePermitWithError(res, permit, paygovError));
+    .then(paygovError => updatePermit(permit, {
+      status: 'Error',
+      paygovError: JSON.stringify(paygovError)
+    })
+      .then(updatedPermit => res.status(400).json(formatPermitError(updatedPermit))));
 };
 
 /**
  * @function grabAndProcessPaygovToken - Private function to error objects from the permit's errors
  * @param {Object} xmlDatafromPayGov - xml from pay.gov
  * @param {Object} permit - permit object from database
  * @param {Object} res - http response
- * @return {Object} - promise of whether permit was updated with token
+ * @return {Promise} - resolves if permit was updated with token
  */
-const grabAndProcessPaygovToken = (payGovXmlRes, permit, res) => new Promise((resolve, reject) => {
-  xml2jsParse(payGovXmlRes, (parseErr, result) => {
-    if (!parseErr) {
-      paygov.getToken(result)
-        .then(token => resolve(updatePermitWithToken(res, permit, token)))
-        .catch(error => recordPayGovError(error, result, res, permit, 'startOnlineCollection'));
-    }
-    reject(parseErr);
-  });
-});
+const grabAndProcessPaygovToken = (payGovXmlRes, permit, res) => util.parseXml(payGovXmlRes)
```

Comment:
Agree 100%. This PR does include one change along those lines, but I felt more refactoring of this type of logic (which exists in many places) should be saved for future PRs.

Reply:
Yeah, that's fair. IMHO, a gnarly bug like this one (involving a third-party service, dealing with credit cards, customer data, multi-step transaction, etc) is a signal to do these kinds of heavier refactors, at least for this one controller. That way it's easier to reason how the fix is correct. For me, it's hard for me to read the code and have confidence that the bug is fixed. The additional tests are good, so that that increases confidence without me actually understanding the code.